src/main/scala/resource/SharedResource.scala

```diff
@@ -0,0 +1,59 @@
+package resource
+
+import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.atomic.AtomicInteger
+
+final class SharedResource[A : Resource : Manifest](opener: => A) extends AbstractManagedResource[A] {
+	private val resource = implicitly[Resource[A]]
+	// A referencing counting atomic variable we use to "guard" this resource.
+	// Currently, we spin-lock accessing this ref because we don't expect opening to block a thread for very long.
+	private val ref = new AtomicReference[Option[A]](None)
+	private val count = new AtomicInteger(0)
+
+	protected def open: A = {
+		val myCount = count.addAndGet(1)
+		// If we win the race, we construct the resource
+		if (myCount == 1) {
+			// We must construct the reference
+			// We may want to CAS here and fail if we're inconsistent.
+			val r = opener
```

Comment:
When one thread failed to open a resource, what is the expected behavior for other tries from other threads?
One possible solution is `synchronized` the entire `open` method, including the `addAndGet`. So that it reset count to zero if fails, and let other threads retry `openr`.